File: packages/ce-core/src/beads.ts
/**
 * BEADS JSONL Format Support
 *
 * Implements reading/writing of the BEADS (Bead-based Engineering and
 * Development System) JSONL format for agent context handoff.
 *
 * BEADS is Steve Yegge's git-backed issue tracker for AI coding agents.
 * The JSONL format is the git-portable serialization layer that enables
 * agents to come and go while maintaining structured context.
 *
 * This module bridges the Context Engineering Toolkit's ContextItem
 * format with BEADS issues, enabling:
 * - Serializing context state to BEADS JSONL for agent handoff
 * - Deserializing BEADS JSONL to pick up where another agent left off
 * - Tracking work items alongside context in a single format
 *
 * Reference: https://github.com/steveyegge/beads
 */

import type { ContextItem, ContextPack } from "./types.js";
import { estimateTokens } from "./estimate.js";

// ─── BEADS Types (subset relevant to context engineering) ────────────

export type BeadsStatus =
  | "open"
  | "in_progress"
  | "blocked"
  | "deferred"
  | "closed"
  | "pinned"
  | "hooked";

export type BeadsIssueType =
  | "bug"
  | "feature"
  | "task"
  | "epic"
  | "chore"
  | "decision"
  | "message"
  | "molecule"
  | "context";  // extension: context items stored as BEADS

export type BeadsDependencyType =
  | "blocks"
  | "parent-child"
  | "related"
  | "replies-to"
  | "supersedes"
  | "caused-by"
  | string;

export interface BeadsDependency {
  issue_id: string;
  depends_on_id: string;
  type: BeadsDependencyType;
  created_at?: string;
  created_by?: string;
}

export interface BeadsComment {
  id?: number;
  issue_id: string;
  author: string;
  text: string;
  created_at: string;
}

/**
 * A BEADS issue — the core record type in the JSONL format.
 *
 * We support the full read schema but only require the subset
 * needed for context engineering operations.
 */
export interface BeadsIssue {
  // Core identification
  id: string;

  // Content
  title: string;
  description?: string;
  design?: string;
  acceptance_criteria?: string;
  notes?: string;

  // Status & workflow
  status: BeadsStatus;
  priority: number;
  issue_type: BeadsIssueType;

  // Assignment
  assignee?: string;
  owner?: string;

  // Timestamps
  created_at: string;
  updated_at: string;
  closed_at?: string;
  close_reason?: string;

  // Scheduling
  due_at?: string;
  defer_until?: string;

  // External
  external_ref?: string;
  source_system?: string;

  // Metadata (arbitrary JSON)
  metadata?: Record<string, unknown>;

  // Compaction
  compaction_level?: number;

  // Relations (populated on export)
  labels?: string[];
  dependencies?: BeadsDependency[];
  comments?: BeadsComment[];

  // Context engineering extensions (stored in metadata)
  // These are convenience fields that map to/from metadata
  pinned?: boolean;
  ephemeral?: boolean;

  // Allow additional fields from the full BEADS spec
  [key: string]: unknown;
}

// ─── BEADS JSONL Read/Write ──────────────────────────────────────────

/**
 * Parse a BEADS JSONL string into an array of issues.
 * Each line is a self-contained JSON object.
 * Blank lines and lines starting with # are skipped.
 *
 * @param input - BEADS JSONL string to parse
 * @returns Array of parsed BEADS issues
 *
 * @example
 * ```ts
 * const jsonl = fs.readFileSync(".beads/issues.jsonl", "utf-8");
 * const issues = readBeadsJSONL(jsonl);
 * console.log(issues.length);
 * ```
 */
export function readBeadsJSONL(input: string): BeadsIssue[] {
  const issues: BeadsIssue[] = [];

  for (const line of input.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;

    try {
      const parsed = JSON.parse(trimmed);
      if (parsed && typeof parsed === "object" && parsed.id) {
        issues.push(parsed as BeadsIssue);
      }
    } catch {
      // Skip malformed lines (BEADS spec tolerates this)
    }
  }

  return issues;
}

/**
 * Serialize an array of BEADS issues to JSONL format.
 * One JSON object per line, no trailing newline.
 *
 * @param issues - Array of BEADS issues to serialize
 * @returns JSONL string with one issue per line
 *
 * @example
 * ```ts
 * const jsonl = writeBeadsJSONL(issues);
 * fs.writeFileSync(".beads/issues.jsonl", jsonl);
 * ```
 */
export function writeBeadsJSONL(issues: BeadsIssue[]): string {
  return issues
    .map(issue => JSON.stringify(issue))
    .join("\n");
}

// ─── ContextItem ↔ BeadsIssue Bridge ─────────────────────────────────

export interface BeadsBridgeOptions {
  /** Agent identity for assignee field */
  agent?: string;
  /** Source system identifier */
  sourceSystem?: string;
  /** Default status for new issues */
  defaultStatus?: BeadsStatus;
  /** Map context item kinds to BEADS issue types */
  kindToIssueType?: Record<string, BeadsIssueType>;
  /** Map context item kinds to BEADS labels */
  kindToLabels?: Record<string, string[]>;
}

const DEFAULT_KIND_TO_ISSUE_TYPE: Record<string, BeadsIssueType> = {
  system: "context",
  tool: "context",
  schema: "context",
  memory: "context",
  conversation: "context",
  query: "context",
  retrieval: "context",
  task: "task",
  bug: "bug",
  feature: "feature",
};

/**
 * Convert a ContextItem to a BEADS issue.
 *
 * The context item's content becomes the description, its kind maps
 * to issue_type and labels, and its priority maps to BEADS priority
 * (inverted: high context priority → low BEADS priority number).
 *
 * @param item - The context item to convert
 * @param options - Bridge options for agent, source system, and kind mapping
 *
 * @example
 * ```ts
 * const issue = contextItemToBeads(
 *   { id: "doc1", content: "API docs...", kind: "system", priority: 10 },
 *   { agent: "agent-1" },
 * );
 * ```
 */
export function contextItemToBeads(
  item: ContextItem,
  options: BeadsBridgeOptions = {},
): BeadsIssue {
  const now = new Date().toISOString();
  const kindMap = options.kindToIssueType ?? DEFAULT_KIND_TO_ISSUE_TYPE;
  const labelMap = options.kindToLabels ?? {};

  // Map priority: ContextItem priority 10 (high) → BEADS P0 (critical)
  // ContextItem priority 1 (low) → BEADS P4 (backlog)
  const beadsPriority = Math.max(0, Math.min(4,
    4 - Math.floor(((item.priority ?? 5) / 10) * 4)
  ));

  const labels: string[] = [];
  if (item.kind) {
    labels.push(`kind:${item.kind}`);
    if (labelMap[item.kind]) {
      labels.push(...labelMap[item.kind]);
    }
  }
  labels.push("context-engineering");

  return {
    id: `ce-${item.id}`,
    title: item.id,
    description: item.content,
    status: options.defaultStatus ?? "open",
    priority: beadsPriority,
    issue_type: kindMap[item.kind ?? ""] ?? "context",
    assignee: options.agent,
    source_system: options.sourceSystem ?? "context-engineering",
    labels,
    created_at: now,
    updated_at: now,
    metadata: {
      ...item.metadata,
      _ce: {
        kind: item.kind,
        priority: item.priority,
        recency: item.recency,
        tokens: item.tokens,
        score: item.score,
        originalId: item.id,
      },
    },
  };
}

/**
 * Convert a BEADS issue back to a ContextItem.
 *
 * Reads the _ce metadata extension to recover original context
 * item properties. Falls back to inferring from BEADS fields.
 *
 * @param issue - The BEADS issue to convert back to a context item
 */
export function beadsToContextItem(issue: BeadsIssue): ContextItem {
  const ceMetadata = (issue.metadata as Record<string, unknown>)?._ce as Record<string, unknown> | undefined;

  // Recover original ID: strip ce- prefix if present
  const originalId = (ceMetadata?.originalId as string) ??
    (issue.id.startsWith("ce-") ? issue.id.slice(3) : issue.id);

  // Recover kind from metadata or labels
  let kind = ceMetadata?.kind as string | undefined;
  if (!kind && issue.labels) {
    const kindLabel = issue.labels.find(l => l.startsWith("kind:"));
    if (kindLabel) kind = kindLabel.slice(5);
  }

  // Recover priority: BEADS P0 → priority 10, P4 → priority 1
  const priority = (ceMetadata?.priority as number) ??
    Math.max(1, Math.round(((4 - issue.priority) / 4) * 10));

  const content = issue.description ?? issue.title;
  const tokens = (ceMetadata?.tokens as number) ?? estimateTokens(content);

  // Strip _ce from metadata to avoid recursion
  const metadata = issue.metadata ? { ...issue.metadata } : undefined;
  if (metadata) delete metadata._ce;

  return {
    id: originalId,
    content,
    kind,
    priority,
    recency: ceMetadata?.recency as number | undefined,
    tokens,
    score: ceMetadata?.score as number | undefined,
    metadata: metadata && Object.keys(metadata).length > 0 ? metadata : undefined,
  };
}

// ─── Agent Handoff Protocol ──────────────────────────────────────────

export interface HandoffOptions extends BeadsBridgeOptions {
  /** Session identifier for traceability */
  sessionId?: string;
  /** Include dropped items as deferred issues */
  includeDropped?: boolean;
  /** Additional notes for the handoff */
  handoffNotes?: string;
}

export interface HandoffResult {
  /** BEADS JSONL string ready to write to .beads/issues.jsonl */
  jsonl: string;
  /** Parsed issues for programmatic access */
  issues: BeadsIssue[];
  /** Summary statistics */
  stats: {
    totalIssues: number;
    contextIssues: number;
    activeItems: number;
    deferredItems: number;
  };
}

/**
 * Create a BEADS JSONL handoff from a context pack.
 *
 * Converts the packed context into BEADS issues that another agent
 * can pick up. Selected items become open issues, dropped items
 * become deferred issues (if includeDropped is true).
 *
 * @param pack - The context pack to convert to a handoff
 * @param options - Handoff options including agent identity and session ID
 * @returns HandoffResult with JSONL string, parsed issues, and stats
 *
 * @example
 * ```ts
 * const pack = session.compile();
 * const handoff = createHandoff(pack, {
 *   agent: "agent-1",
 *   sessionId: "session-abc",
 *   includeDropped: true,
 * });
 * fs.writeFileSync(".beads/issues.jsonl", handoff.jsonl);
 * // git add . && git push
 * ```
 */
export function createHandoff(
  pack: ContextPack,
  options: HandoffOptions = {},
): HandoffResult {
  const issues: BeadsIssue[] = [];
  const now = new Date().toISOString();

  // Create a manifest issue (the "handoff bead")
  const manifestIssue: BeadsIssue = {
    id: `ce-handoff-${Date.now().toString(36)}`,
    title: "Context Engineering Handoff",
    description: options.handoffNotes ?? "Agent context handoff via Context Engineering Toolkit",
    status: "pinned",
    priority: 0,
    issue_type: "message",
    assignee: options.agent,
    source_system: options.sourceSystem ?? "context-engineering",
    labels: ["context-engineering", "handoff"],
    created_at: now,
    updated_at: now,
    metadata: {
      _ce_handoff: {
        sessionId: options.sessionId,
        totalTokens: pack.totalTokens,
        selectedCount: pack.selected.length,
        droppedCount: pack.dropped.length,
        budget: pack.budget,
        stats: pack.stats,
        createdAt: now,
      },
    },
  };
  issues.push(manifestIssue);

  // Convert selected items to open issues
  for (const item of pack.selected) {
    const issue = contextItemToBeads(item, options);
    issue.status = "open";
    issues.push(issue);
  }

  // Convert dropped items to deferred issues
  if (options.includeDropped) {
    for (const item of pack.dropped) {
      const issue = contextItemToBeads(item, options);
      issue.status = "deferred";
      issue.defer_until = now; // immediately available on pickup
      issues.push(issue);
    }
  }

  return {
    jsonl: writeBeadsJSONL(issues),
    issues,
    stats: {
      totalIssues: issues.length,
      contextIssues: issues.length - 1, // exclude manifest
      activeItems: pack.selected.length,
      deferredItems: options.includeDropped ? pack.dropped.length : 0,
    },
  };
}

export interface PickupResult {
  /** Recovered context items (from open issues) */
  items: ContextItem[];
  /** Deferred context items (from deferred issues) */
  deferred: ContextItem[];
  /** The handoff manifest, if found */
  manifest: BeadsIssue | null;
  /** Non-context issues (tasks, bugs, etc.) */
  workItems: BeadsIssue[];
  /** Summary statistics */
  stats: {
    totalIssues: number;
    contextItems: number;
    deferredItems: number;
    workItems: number;
    handoffSessionId?: string;
    handoffBudget?: unknown;
  };
}

/**
 * Pick up context from a BEADS JSONL handoff.
 *
 * Reads the JSONL, separates context items from work items,
 * and recovers the original ContextItem format.
 *
 * @param jsonl - BEADS JSONL string to parse
 * @returns PickupResult with recovered context items, deferred items, and work items
 *
 * @example
 * ```ts
 * const jsonl = fs.readFileSync(".beads/issues.jsonl", "utf-8");
 * const pickup = pickupHandoff(jsonl);
 *
 * const session = createSession({ budget: { maxTokens: 8000 } });
 * session.setItems(pickup.items);
 * const result = session.compile();
 * ```
 */
export function pickupHandoff(jsonl: string): PickupResult {
  const issues = readBeadsJSONL(jsonl);

  const items: ContextItem[] = [];
  const deferred: ContextItem[] = [];
  const workItems: BeadsIssue[] = [];
  let manifest: BeadsIssue | null = null;

  for (const issue of issues) {
    // Check if it's a handoff manifest
    if (
      issue.id.startsWith("ce-handoff-") ||
      (issue.labels?.includes("handoff") && issue.issue_type === "message")
    ) {
      manifest = issue;
      continue;
    }

    // Check if it's a context engineering item
    const isContextItem =
      issue.issue_type === "context" ||
      issue.source_system === "context-engineering" ||
      issue.labels?.includes("context-engineering") ||
      (issue.metadata as Record<string, unknown>)?._ce !== undefined;

    if (isContextItem) {
      const contextItem = beadsToContextItem(issue);
      if (issue.status === "deferred") {
        deferred.push(contextItem);
      } else if (issue.status !== "closed") {
        items.push(contextItem);
      }
    } else {
      workItems.push(issue);
    }
  }

  const handoffMeta = (manifest?.metadata as Record<string, unknown>)?._ce_handoff as Record<string, unknown> | undefined;

  return {
    items,
    deferred,
    manifest,
    workItems,
    stats: {
      totalIssues: issues.length,
      contextItems: items.length,
      deferredItems: deferred.length,
      workItems: workItems.length,
      handoffSessionId: handoffMeta?.sessionId as string | undefined,
      handoffBudget: handoffMeta?.budget,
    },
  };
}

// ─── Incremental BEADS Operations ────────────────────────────────────

/**
 * Merge new issues into an existing BEADS JSONL string.
 * Issues with the same ID are replaced; new issues are appended.
 *
 * @param existing - Existing BEADS JSONL string
 * @param updates - New or updated issues to merge in
 * @returns Merged JSONL string
 */
export function mergeBeadsJSONL(
  existing: string,
  updates: BeadsIssue[],
): string {
  const existingIssues = readBeadsJSONL(existing);
  const existingMap = new Map(existingIssues.map(i => [i.id, i]));

  for (const update of updates) {
    existingMap.set(update.id, update);
  }

  return writeBeadsJSONL(Array.from(existingMap.values()));
}

/**
 * Filter BEADS issues to get only "ready" items — the equivalent
 * of `bd ready`. Returns issues that are:
 * - Status is "open"
 * - Not ephemeral
 * - No unresolved blocking dependencies
 * - defer_until is null or in the past
 *
 * @param issues - Array of BEADS issues to filter
 * @returns Issues that are ready to be worked on
 */
export function getReadyIssues(issues: BeadsIssue[]): BeadsIssue[] {
  const now = new Date().toISOString();
  const closedIds = new Set(
    issues.filter(i => i.status === "closed").map(i => i.id)
  );

  return issues.filter(issue => {
    if (issue.status !== "open") return false;
    if (issue.ephemeral) return false;

    // Check defer_until
    if (issue.defer_until && issue.defer_until > now) return false;

    // Check blocking dependencies
    if (issue.dependencies) {
      const blocked = issue.dependencies.some(dep => {
        if (dep.type !== "blocks" && dep.type !== "parent-child") return false;
        return !closedIds.has(dep.depends_on_id);
      });
      if (blocked) return false;
    }

    return true;
  });
}

File: python/context_engineering/beads.py
"""
BEADS JSONL Format Support

Implements reading/writing of the BEADS JSONL format for agent context handoff.

BEADS is Steve Yegge's git-backed issue tracker for AI coding agents.
The JSONL format is the git-portable serialization layer that enables
agents to come and go while maintaining structured context.

This module bridges the Context Engineering Toolkit's ContextItem
format with BEADS issues, enabling:
- Serializing context state to BEADS JSONL for agent handoff
- Deserializing BEADS JSONL to pick up where another agent left off
- Tracking work items alongside context in a single format

Reference: https://github.com/steveyegge/beads
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Literal, Optional

from .core import ContextItem, ContextPack, estimate_tokens

# ─── BEADS Types ──────────────────────────────────────────────────────

BeadsStatus = Literal["open", "in_progress", "blocked", "deferred", "closed", "pinned", "hooked"]

BeadsIssueType = Literal[
    "bug", "feature", "task", "epic", "chore", "decision", "message", "molecule", "context"
]


@dataclass
class BeadsDependency:
    issue_id: str
    depends_on_id: str
    type: str  # "blocks", "parent-child", "related", etc.
    created_at: Optional[str] = None
    created_by: Optional[str] = None


@dataclass
class BeadsComment:
    issue_id: str
    author: str
    text: str
    created_at: str
    id: Optional[int] = None


@dataclass
class BeadsIssue:
    """A BEADS issue — the core record type in the JSONL format."""

    id: str
    title: str
    status: str  # BeadsStatus
    priority: int
    issue_type: str  # BeadsIssueType
    created_at: str
    updated_at: str

    description: Optional[str] = None
    design: Optional[str] = None
    acceptance_criteria: Optional[str] = None
    notes: Optional[str] = None

    assignee: Optional[str] = None
    owner: Optional[str] = None

    closed_at: Optional[str] = None
    close_reason: Optional[str] = None
    due_at: Optional[str] = None
    defer_until: Optional[str] = None

    external_ref: Optional[str] = None
    source_system: Optional[str] = None

    metadata: Optional[Dict[str, Any]] = None
    compaction_level: Optional[int] = None

    labels: Optional[List[str]] = None
    dependencies: Optional[List[BeadsDependency]] = None
    comments: Optional[List[BeadsComment]] = None

    pinned: Optional[bool] = None
    ephemeral: Optional[bool] = None

    # Catch-all for additional BEADS fields
    extra: Optional[Dict[str, Any]] = field(default=None, repr=False)

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to a dict, excluding None values."""
        d: Dict[str, Any] = {}
        for k, v in self.__dict__.items():
            if v is None or k == "extra":
                continue
            if k == "dependencies" and v:
                d[k] = [{kk: vv for kk, vv in dep.__dict__.items() if vv is not None} for dep in v]
            elif k == "comments" and v:
                d[k] = [{kk: vv for kk, vv in c.__dict__.items() if vv is not None} for c in v]
            else:
                d[k] = v
        if self.extra:
            d.update(self.extra)
        return d

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "BeadsIssue":
        """Deserialize from a dict, storing unknown fields in extra."""
        known_fields = {f for f in cls.__dataclass_fields__}
        known_data = {}
        extra_data = {}

        for k, v in data.items():
            if k in known_fields:
                known_data[k] = v
            else:
                extra_data[k] = v

        # Parse dependencies
        if "dependencies" in known_data and known_data["dependencies"]:
            known_data["dependencies"] = [
                BeadsDependency(**dep) if isinstance(dep, dict) else dep
                for dep in known_data["dependencies"]
            ]

        # Parse comments
        if "comments" in known_data and known_data["comments"]:
            known_data["comments"] = [
                BeadsComment(**c) if isinstance(c, dict) else c for c in known_data["comments"]
            ]

        issue = cls(**known_data)
        if extra_data:
            issue.extra = extra_data
        return issue


# ─── BEADS JSONL Read/Write ───────────────────────────────────────────


def read_beads_jsonl(input_str: str) -> List[BeadsIssue]:
    """Parse a BEADS JSONL string into a list of issues.

    Each line is a self-contained JSON object.
    Blank lines and lines starting with # are skipped.
    """
    issues: List[BeadsIssue] = []

    for line in input_str.split("\n"):
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue

        try:
            parsed = json.loads(trimmed)
            if isinstance(parsed, dict) and "id" in parsed:
                issues.append(BeadsIssue.from_dict(parsed))
        except (json.JSONDecodeError, TypeError):
            pass  # Skip malformed lines

    return issues


def write_beads_jsonl(issues: List[BeadsIssue]) -> str:
    """Serialize a list of BEADS issues to JSONL format."""
    return "\n".join(json.dumps(issue.to_dict()) for issue in issues)


# ─── ContextItem ↔ BeadsIssue Bridge ─────────────────────────────────


@dataclass
class BeadsBridgeOptions:
    """Options for bridging between ContextItems and BEADS issues."""

    agent: Optional[str] = None
    source_system: str = "context-engineering"
    default_status: str = "open"
    kind_to_issue_type: Optional[Dict[str, str]] = None
    kind_to_labels: Optional[Dict[str, List[str]]] = None


_DEFAULT_KIND_TO_ISSUE_TYPE: Dict[str, str] = {
    "system": "context",
    "tool": "context",
    "schema": "context",
    "memory": "context",
    "conversation": "context",
    "query": "context",
    "retrieval": "context",
    "task": "task",
    "bug": "bug",
    "feature": "feature",
}


def context_item_to_beads(
    item: ContextItem,
    options: Optional[BeadsBridgeOptions] = None,
) -> BeadsIssue:
    """Convert a ContextItem to a BEADS issue.

    The context item's content becomes the description, its kind maps
    to issue_type and labels, and its priority maps to BEADS priority
    (inverted: high context priority -> low BEADS priority number).
    """
    opts = options or BeadsBridgeOptions()
    now = datetime.now(timezone.utc).isoformat()
    kind_map = opts.kind_to_issue_type or _DEFAULT_KIND_TO_ISSUE_TYPE
    label_map = opts.kind_to_labels or {}

    # Map priority: ContextItem priority 10 (high) -> BEADS P0 (critical)
    priority_val = item.priority if item.priority is not None else 5.0
    beads_priority = max(0, min(4, 4 - int((priority_val / 10) * 4)))

    labels = []
    if item.kind:
        labels.append(f"kind:{item.kind}")
        if item.kind in label_map:
            labels.extend(label_map[item.kind])
    labels.append("context-engineering")

    metadata: Dict[str, Any] = {**(item.metadata or {})}
    metadata["_ce"] = {
        "kind": item.kind,
        "priority": item.priority,
        "recency": item.recency,
        "tokens": item.tokens,
        "score": item.score,
        "originalId": item.id,
    }

    return BeadsIssue(
        id=f"ce-{item.id}",
        title=item.id,
        description=item.content,
        status=opts.default_status,
        priority=beads_priority,
        issue_type=kind_map.get(item.kind or "", "context"),
        assignee=opts.agent,
        source_system=opts.source_system,
        labels=labels,
        created_at=now,
        updated_at=now,
        metadata=metadata,
    )


def beads_to_context_item(issue: BeadsIssue) -> ContextItem:
    """Convert a BEADS issue back to a ContextItem.

    Reads the _ce metadata extension to recover original context
    item properties. Falls back to inferring from BEADS fields.
    """
    ce_metadata = (issue.metadata or {}).get("_ce", {}) if issue.metadata else {}

    # Recover original ID
    original_id = ce_metadata.get("originalId")
    if not original_id:
        original_id = issue.id[3:] if issue.id.startswith("ce-") else issue.id

    # Recover kind from metadata or labels
    kind = ce_metadata.get("kind")
    if not kind and issue.labels:
        for label in issue.labels:
            if label.startswith("kind:"):
                kind = label[5:]
                break

    # Recover priority
    priority = ce_metadata.get("priority")
    if priority is None:
        priority = max(1, round(((4 - issue.priority) / 4) * 10))

    content = issue.description or issue.title
    tokens = ce_metadata.get("tokens") or estimate_tokens(content)

    # Strip _ce from metadata
    metadata = dict(issue.metadata) if issue.metadata else {}
    metadata.pop("_ce", None)

    return ContextItem(
        id=original_id,
        content=content,
        kind=kind,
        priority=priority,
        recency=ce_metadata.get("recency"),
        tokens=tokens,
        score=ce_metadata.get("score"),
        metadata=metadata if metadata else {},
    )


# ─── Agent Handoff Protocol ──────────────────────────────────────────


@dataclass
class HandoffOptions(BeadsBridgeOptions):
    """Options for creating a handoff."""

    session_id: Optional[str] = None
    include_dropped: bool = False
    handoff_notes: Optional[str] = None


@dataclass
class HandoffResult:
    """Result of creating a BEADS handoff."""

    jsonl: str
    issues: List[BeadsIssue]
    stats: Dict[str, Any]


def create_handoff(
    pack: ContextPack,
    options: Optional[HandoffOptions] = None,
) -> HandoffResult:
    """Create a BEADS JSONL handoff from a context pack.

    Converts the packed context into BEADS issues that another agent
    can pick up. Selected items become open issues, dropped items
    become deferred issues (if include_dropped is True).

    Example:
        pack = session.compile()
        handoff = create_handoff(pack, HandoffOptions(
            agent="agent-1",
            session_id="session-abc",
            include_dropped=True,
        ))
        with open(".beads/issues.jsonl", "w") as f:
            f.write(handoff.jsonl)
    """
    opts = options or HandoffOptions()
    now = datetime.now(timezone.utc).isoformat()
    issues: List[BeadsIssue] = []

    # Create manifest issue (the "handoff bead")
    manifest = BeadsIssue(
        id=f"ce-handoff-{hex(int(time.time() * 1000))[2:]}",
        title="Context Engineering Handoff",
        description=opts.handoff_notes or "Agent context handoff via Context Engineering Toolkit",
        status="pinned",
        priority=0,
        issue_type="message",
        assignee=opts.agent,
        source_system=opts.source_system,
        labels=["context-engineering", "handoff"],
        created_at=now,
        updated_at=now,
        metadata={
            "_ce_handoff": {
                "sessionId": opts.session_id,
                "totalTokens": pack.total_tokens,
                "selectedCount": len(pack.selected),
                "droppedCount": len(pack.dropped),
                "budget": {
                    "maxTokens": pack.budget.max_tokens,
                    "reserveTokens": pack.budget.reserve_tokens,
                },
                "createdAt": now,
            },
        },
    )
    issues.append(manifest)

    # Convert selected items to open issues
    bridge_opts = BeadsBridgeOptions(
        agent=opts.agent,
        source_system=opts.source_system,
        default_status="open",
        kind_to_issue_type=opts.kind_to_issue_type,
        kind_to_labels=opts.kind_to_labels,
    )
    for item in pack.selected:
        issue = context_item_to_beads(item, bridge_opts)
        issue.status = "open"
        issues.append(issue)

    # Convert dropped items to deferred issues
    deferred_count = 0
    if opts.include_dropped:
        for item in pack.dropped:
            issue = context_item_to_beads(item, bridge_opts)
            issue.status = "deferred"
            issue.defer_until = now
            issues.append(issue)
            deferred_count += 1

    return HandoffResult(
        jsonl=write_beads_jsonl(issues),
        issues=issues,
        stats={
            "totalIssues": len(issues),
            "contextIssues": len(issues) - 1,
            "activeItems": len(pack.selected),
            "deferredItems": deferred_count,
        },
    )


@dataclass
class PickupResult:
    """Result of picking up context from a BEADS handoff."""

    items: List[ContextItem]
    deferred: List[ContextItem]
    manifest: Optional[BeadsIssue]
    work_items: List[BeadsIssue]
    stats: Dict[str, Any]


def pickup_handoff(jsonl: str) -> PickupResult:
    """Pick up context from a BEADS JSONL handoff.

    Reads the JSONL, separates context items from work items,
    and recovers the original ContextItem format.

    Example:
        with open(".beads/issues.jsonl") as f:
            jsonl = f.read()
        pickup = pickup_handoff(jsonl)

        session = create_session(budget)
        session.set_items(pickup.items)
        result = session.compile()
    """
    issues = read_beads_jsonl(jsonl)

    items: List[ContextItem] = []
    deferred: List[ContextItem] = []
    work_items: List[BeadsIssue] = []
    manifest: Optional[BeadsIssue] = None

    for issue in issues:
        # Check if it's a handoff manifest
        if issue.id.startswith("ce-handoff-") or (
            issue.labels and "handoff" in issue.labels and issue.issue_type == "message"
        ):
            manifest = issue
            continue

        # Check if it's a context engineering item
        is_context = (
            issue.issue_type == "context"
            or issue.source_system == "context-engineering"
            or (issue.labels and "context-engineering" in issue.labels)
            or (issue.metadata and "_ce" in issue.metadata)
        )

        if is_context:
            context_item = beads_to_context_item(issue)
            if issue.status == "deferred":
                deferred.append(context_item)
            elif issue.status != "closed":
                items.append(context_item)
        else:
            work_items.append(issue)

    handoff_meta = {}
    if manifest and manifest.metadata:
        handoff_meta = manifest.metadata.get("_ce_handoff", {})

    return PickupResult(
        items=items,
        deferred=deferred,
        manifest=manifest,
        work_items=work_items,
        stats={
            "totalIssues": len(issues),
            "contextItems": len(items),
            "deferredItems": len(deferred),
            "workItems": len(work_items),
            "handoffSessionId": handoff_meta.get("sessionId"),
            "handoffBudget": handoff_meta.get("budget"),
        },
    )


# ─── Incremental BEADS Operations ────────────────────────────────────


def merge_beads_jsonl(existing: str, updates: List[BeadsIssue]) -> str:
    """Merge new issues into an existing BEADS JSONL string.

    Issues with the same ID are replaced; new issues are appended.
    """
    existing_issues = read_beads_jsonl(existing)
    existing_map = {i.id: i for i in existing_issues}

    for update in updates:
        existing_map[update.id] = update

    return write_beads_jsonl(list(existing_map.values()))


def get_ready_issues(issues: List[BeadsIssue]) -> List[BeadsIssue]:
    """Filter BEADS issues to get only "ready" items.

    Equivalent to `bd ready`. Returns issues that are:
    - Status is "open"
    - Not ephemeral
    - No unresolved blocking dependencies
    - defer_until is null or in the past
    """
    now = datetime.now(timezone.utc).isoformat()
    closed_ids = {i.id for i in issues if i.status == "closed"}

    ready = []
    for issue in issues:
        if issue.status != "open":
            continue
        if issue.ephemeral:
            continue
        if issue.defer_until and issue.defer_until > now:
            continue

        # Check blocking dependencies
        if issue.dependencies:
            blocked = False
            for dep in issue.dependencies:
                if dep.type in ("blocks", "parent-child"):
                    if dep.depends_on_id not in closed_ids:
                        blocked = True
                        break
            if blocked:
                continue

        ready.append(issue)

    return ready
